File: C_Sharp/Framework/LazyLoad/LazyLoadList.cs
using System.Collections.Generic;
using Framework.Domain;

namespace Framework.LazyLoad
{
    public class LazyLoadList<TEntity, TSearchInput> : List<TEntity>
        where TEntity : LazyLoadDomainObject<TSearchInput>, IDomainObject
    {
        private ILazyLoader<TEntity, TSearchInput> _loader;

        public LazyLoadList(ILazyLoader<TEntity, TSearchInput> loader)
        {
            _loader = loader;
        }

        public TEntity this[int index]
        {
            get
            {
                TEntity entity = base[index];

                if ((entity == null) || (entity.Criteria == null) || (_loader == null))
                    return entity;

                //Perform complete load
                entity.Load(_loader, entity.Criteria);

                return entity;
            }
        }
    }
}

File: Android/framework/src/main/java/com/LazyLoad/LazyLoadList.java
package com.LazyLoad;

import com.Domain.Interfaces.IDomainObject;
import com.Interfaces.ILazyLoader;

import java.util.ArrayList;

/**
 * Created by aiko on 4/2/17.
 */

public class LazyLoadList<
        TEntity extends LazyLoadDomainObject<TSearchInput>,
        TSearchInput> extends ArrayList<TEntity> {

    ILazyLoader<TEntity, TSearchInput> _loader;

    public LazyLoadList(ILazyLoader<TEntity, TSearchInput> loader)
    {
        _loader = loader;
    }

    @Override
    public TEntity get(int index)
    {
        TEntity entity = super.get(index);

        if ((entity == null) || (_loader == null))
            return entity;

        //Perform complete load
        entity.Load(_loader);

        return entity;
    }
}
